File: src/test/java/uk/co/reecedunn/intellij/plugin/core/tests/vfs/ResourceVirtualFile.java
/*
 * Copyright (C) 2016 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.core.tests.vfs;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileSystem;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URISyntaxException;
import java.net.URL;

public class ResourceVirtualFile extends VirtualFile {
    private ClassLoader mLoader;
    private String mResource;
    private String mParent;
    private String mName;
    private File mFile = null;

    public ResourceVirtualFile(ClassLoader loader, String resource) {
        mLoader = loader;
        mResource = resource;

        int idx = resource.lastIndexOf('/');
        if (idx == -1) {
            mParent = null;
            mName = resource;
        } else {
            mParent = resource.substring(0, idx);
            mName = resource.substring(idx + 1);
        }

        URL url = mLoader.getResource(resource);
        if (url != null && url.getProtocol().equals("file")) {
            try {
                mFile = new File(url.toURI());
            } catch (URISyntaxException e) {
                //
            }
        }
    }

    public ResourceVirtualFile(String resource) {
        this(ResourceVirtualFile.class.getClassLoader(), resource);
    }

    @NotNull
    @Override
    public String getName() {
        return mName;
    }

    @NotNull
    @Override
    public VirtualFileSystem getFileSystem() {
        throw new UnsupportedOperationException();
    }

    @NotNull
    @Override
    public String getPath() {
        return mFile != null ? mFile.getPath() : "";
    }

    @Override
    public boolean isWritable() {
        return false;
    }

    @Override
    public boolean isDirectory() {
        return mFile != null && mFile.isDirectory();
    }

    @Override
    public boolean isValid() {
        return mFile != null;
    }

    @Override
    public VirtualFile getParent() {
        return mParent == null ? null : new ResourceVirtualFile(mLoader, mParent);
    }

    @Override
    public VirtualFile[] getChildren() {
        if (mFile == null) {
            return null;
        }

        String[] children = mFile.list();
        if (children == null) {
            return null;
        }

        VirtualFile[] files = new VirtualFile[children.length];
        for (int i = 0; i != children.length; ++i) {
            files[i] = new ResourceVirtualFile(mLoader, mResource + "/" + children[i]);
        }
        return files;
    }

    @NotNull
    @Override
    public OutputStream getOutputStream(Object requestor, long newModificationStamp, long newTimeStamp) throws IOException {
        throw new UnsupportedOperationException();
    }

    @NotNull
    @Override
    public byte[] contentsToByteArray() throws IOException {
        throw new UnsupportedOperationException();
    }

    @Override
    public long getTimeStamp() {
        throw new UnsupportedOperationException();
    }

    @Override
    public long getLength() {
        throw new UnsupportedOperationException();
    }

    @Override
    public void refresh(boolean asynchronous, boolean recursive, @Nullable Runnable postRunnable) {
        throw new UnsupportedOperationException();
    }

    @Override
    public InputStream getInputStream() throws IOException {
        return isDirectory() ? null : mLoader.getResourceAsStream(mResource);
    }
}

File: src/lang-core/main/uk/co/reecedunn/intellij/plugin/core/vfs/ResourceVirtualFile.kt
/*
 * Copyright (C) 2016, 2018-2019 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.core.vfs

import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileSystem
import org.apache.commons.compress.utils.IOUtils

import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.JarURLConnection

class ResourceVirtualFile private constructor(
    private val mLoader: ClassLoader,
    private val mResource: String,
    private val mFileSystem: VirtualFileSystem?,
    private var mFile: File?,
    private var mPath: String?
) : VirtualFile() {
    private var mParent: String? = null
    private var mName: String? = null

    init {
        val idx = mResource.lastIndexOf('/')
        if (idx == -1) {
            mParent = null
            mName = mResource
        } else {
            mParent = mResource.substring(0, idx)
            mName = mResource.substring(idx + 1)
        }
    }

    override fun getName(): String = mName!!

    override fun getFileSystem(): VirtualFileSystem = mFileSystem ?: throw UnsupportedOperationException()

    override fun getPath(): String = mPath ?: ""

    override fun isWritable(): Boolean = false

    override fun isDirectory(): Boolean = mFile != null && mFile!!.isDirectory

    override fun isValid(): Boolean = mFile != null

    override fun getParent(): VirtualFile? {
        return if (mParent == null) null else create(mLoader, mParent!!, mFileSystem)
    }

    override fun getChildren(): Array<VirtualFile>? {
        if (mFile == null) {
            return null
        }

        val children = mFile!!.list() ?: return null
        return children.map { child -> create(mLoader, "$mResource/$child", mFileSystem) }.toTypedArray()
    }

    @Throws(IOException::class)
    override fun getOutputStream(requestor: Any, newModificationStamp: Long, newTimeStamp: Long): OutputStream {
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun contentsToByteArray(): ByteArray {
        val input = inputStream ?: throw UnsupportedOperationException()
        return IOUtils.toByteArray(input)
    }

    override fun getTimeStamp(): Long = mFile!!.lastModified()

    override fun getModificationStamp(): Long = 0

    override fun getLength(): Long {
        return if (mFile == null || mFile!!.isDirectory) 0 else mFile!!.length()
    }

    override fun refresh(asynchronous: Boolean, recursive: Boolean, postRunnable: Runnable?) = TODO()

    @Throws(IOException::class)
    override fun getInputStream(): InputStream? {
        return if (isDirectory) null else mLoader.getResourceAsStream(mResource)
    }

    companion object {
        fun create(loader: ClassLoader, resource: String, fileSystem: VirtualFileSystem? = null): ResourceVirtualFile {
            return createIfValid(loader, resource, fileSystem)
                ?: ResourceVirtualFile(loader, resource, fileSystem, null, null)
        }

        fun createIfValid(
            loader: ClassLoader,
            resource: String,
            fileSystem: VirtualFileSystem? = null
        ): ResourceVirtualFile? {
            return loader.getResource(resource)?.let {
                when (it.protocol) {
                    "file" -> {
                        val file = File(it.toURI())
                        ResourceVirtualFile(loader, resource, fileSystem, file, file.path)
                    }
                    "jar" -> {
                        val connection = it.openConnection() as JarURLConnection
                        val file = File(connection.jarFileURL.toURI())
                        ResourceVirtualFile(loader, resource, fileSystem, file, "${file.path}!/$resource")
                    }
                    else -> null
                }
            }
        }
    }
}
